File: 0020-Valid-Parentheses/lc0020.js
// JavaScript solution.

/**
 * @param {string} s
 * @return {boolean}
 */

var isValid = function(s) {
    let st = [];
    for (let c of [...s]) {
        if (c === '{') st.push('}');
        else if (c === '(') st.push(')');
        else if (c === '[') st.push(']');
        else {
            if (st.length === 0 || c !== st.pop()) return false;
        }
    }
    return st.length === 0;
};

File: 0020-Valid-Parentheses/lc0020.py
# Python solution.

class Solution:
    def isValid(self, s: str) -> bool:
        st = []
        matchmap = {')':'(', '}':'{', ']':'['}
        for c in s:
            if c in matchmap:
                if len(st) == 0: return False
                if st.pop() != matchmap[c]: return False
            else:
                st.append(c)
        return st == []
